lib/spack/spack/binary_distribution.py

```diff
@@ -1478,63 +1584,116 @@ def download_tarball(spec, unsigned=False, mirrors_for_spec=None):
-                    #     the remaining mirrors, looking for one we can use.
-                    tarball_stage = try_fetch(spackfile_url)
-                    if tarball_stage:
-                        return {
-                            "tarball_stage": tarball_stage,
-                            "specfile_stage": local_specfile_stage,
-                            "signature_verified": signature_verified,
-                        }
+    for try_signed in (True, False):
+        for mirror in mirrors:
+            # If it's an OCI index, do things differently, since we cannot compose URLs.
+            parsed = urllib.parse.urlparse(mirror)
 
-                local_specfile_stage.destroy()
+            # TODO: refactor this to some "nice" place.
+            if parsed.scheme == "oci":
+                ref = spack.oci.image.ImageReference.from_string(mirror[len("oci://") :]).with_tag(
+                    spack.oci.image.default_tag(spec)
+                )
+
+                # Fetch the manifest
+                try:
+                    response = spack.oci.opener.urlopen(
+                        urllib.request.Request(
+                            url=ref.manifest_url(),
+                            headers={"Accept": "application/vnd.oci.image.manifest.v1+json"},
+                        )
+                    )
+                except Exception:
+                    continue
+
+                # Download the config = spec.json and the relevant tarball
+                try:
+                    manifest = json.loads(response.read())
+                    spec_digest = spack.oci.image.Digest.from_string(manifest["config"]["digest"])
+                    tarball_digest = spack.oci.image.Digest.from_string(
+                        manifest["layers"][-1]["digest"]
+                    )
+                except Exception:
+                    continue
+
+                with spack.oci.oci.make_stage(
+                    ref.blob_url(spec_digest), spec_digest, keep=True
+                ) as local_specfile_stage:
+                    try:
+                        local_specfile_stage.fetch()
+                        local_specfile_stage.check()
+                    except Exception:
+                        continue
+                    local_specfile_stage.cache_local()
+
+                with spack.oci.oci.make_stage(
+                    ref.blob_url(tarball_digest), tarball_digest, keep=True
+                ) as tarball_stage:
+                    try:
+                        tarball_stage.fetch()
+                        tarball_stage.check()
+                    except Exception:
+                        continue
+                    tarball_stage.cache_local()
+
+                return {
+                    "tarball_stage": tarball_stage,
+                    "specfile_stage": local_specfile_stage,
+                    "signature_verified": False,
+                }
+
+            else:
```

Comment:
I'm thinking it would be easier to have them share code if they're more structurally similar, which requires an overhaul of the buildcache file structure.

I ran into a race condition in Gitlab CI, where two PRs concretized to the same spec, but produced a different tarball (so different shasum); the tarballs were uploaded from PR 1 first, then PR 2, whereas spec.json was uploaded from PR 2 first, then PR 1, so that spec.json listed the wrong shasum.

That can be avoided by actually making standard buildcache tarballs content-adressable, which mimics OCI structure. Then in the same situation as above we'd end up with two uploaded tarballs (one "dangling"), and a single (overwritten) spec.json, always listing a valid tarball shasum.

The assumption being that two builds of the same dag hash always produce an equivalent (but not necessarily bit-wise identical) tarball.
 
So, if buildcaches work like that, their (spec.json, tarball) pairs are pretty close to (manifest, layer) pairs from the OCI spec.

Reply:
Decided in meeting that we should get OCI buildcache mostly as-is initially, then migrate the existing buildcache format to be content-addressed.  So this can be a nice refactor later on, but doesn't prevent getting this in now.